### ISSUE TITLE
File extension for download audio in opus coded should be .ogg

### ISSUE THREAD

**Issue by Tanishq-Banyal** [TeamNewPipe/NewPipe#8046]
### Checklist

- [X] I am able to reproduce the bug with the [latest version](https://github.com/TeamNewPipe/NewPipe/releases/latest).
- [X] I made sure that there are *no existing issues* - [open](https://github.com/TeamNewPipe/NewPipe/issues) or [closed](https://github.com/TeamNewPipe/NewPipe/issues?q=is%3Aissue+is%3Aclosed) - which I could contribute my information to.
- [X] I have taken the time to fill in all the required details. I understand that the bug report will be dismissed otherwise.
- [X] This issue contains only one bug.
- [X] I have read and understood the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/dev/.github/CONTRIBUTING.md).

### Affected version

0.22.1

### Steps to reproduce the bug

1. Open any Video
2. Tap on Download Button
3. Select Audio Download in opus codec (any bitrate)

### Expected behavior

File should be downloaded with .ogg extension since even though the codec is opus but the container is actually ogg.

### Actual behavior

File is downloaded with .opus extension, which is not recognised by many apps including Android System itself.

### Affected Android/Custom ROM version

LineageOS 16 (android 9, rooted)

### Affected device model

Xiaomi Redmi 3s Prime (land)

**Comment by TacoTheDank** [comment]
I don't think this is necessarily an issue with Opus as it is with [Android support](https://en.wikipedia.org/wiki/Opus_(audio_format)#:~:text=Google%20added%20native%20support%20for%20Opus%20audio%20playback%20in%20Android%205.0%20%22Lollipop%22.) over different OS versions.

It would be good to consider what we would want the new behavior to be if this were to be changed. Would we want the audio to be saved as `.ogg` for Android 9 or lower, and be saved as `.opus` for Android 10 and above?

**Comment by Tanishq-Banyal** [comment]
Just save with .ogg extension for all android versions because there is no such thing as a .opus file, opus is a codec. The file container is actually a ogg, so correct extension is .ogg

**Comment by opusforlife2** [comment]
>Just save with .ogg extension for all android versions because there is no such thing as a .opus file, opus is a codec. The file container is actually a ogg, so correct extension is .ogg

**Comment by opusforlife2** [comment]
> Just save with .ogg extension for all android versions because there is no such thing as a .opus file, opus is a codec. The file container is actually a ogg, so correct extension is .ogg

@Tanishq-Banyal Your information is incorrect. You're mixing containers and file extensions: https://en.wikipedia.org/wiki/Opus_(audio_format)#cite_note-MIME-and-FileExt-1

Following that link: https://wiki.xiph.org/MIME_Types_and_File_Extensions#.opus_-_audio.2Fogg

So the codec is Opus, and the container format is indeed Ogg, but the "official" file extension is `.opus`. 3 different things.

As a much clearer example, look at MP4: https://en.wikipedia.org/wiki/MPEG-4_Part_14#Filename_extensions:

> While the only filename extension defined by the standard is .mp4, various filename extensions are commonly used to indicate intended content

---

Coming to your issue:

> which is not recognised by many apps including Android System itself

This is the bug you've reported. But LineageOS added support for the `.opus` file extension in Android 9 (LOS 16), the version you've mentioned! See https://review.lineageos.org/q/opus+extension

I've been storing and playing `.opus` files myself since LOS 16. If you use a file explorer app that supports showing `.opus` files, then you can play them just fine. Try using FX File Explorer and Vanilla Music. Both showed and played the files for me.

**Comment by opusforlife2** [comment]
### Part 2 (the comment was getting too long):

> Would we want the audio to be saved as .ogg for Android 9 or lower, and be saved as .opus for Android 10 and above?

@TacoTheDank Using the `.ogg` file extension will only work for Android 7, 8, and 9, because support for it was added in 7: https://en.wikipedia.org/wiki/Opus_(audio_format)#Operating_system_support

Reading that section further, if you want Android System to natively recognize and play Opus audios for Android 5 and 6, you will need the Matroska container with the Opus codec, which was the only combo supported for those versions. You could use the `.mka` file extension instead of `.mkv` to indicate audio content, but I don't know how many file explorers and audio players support _that_.

---

Extra info for the interested: Most audio players only care about the container format, not the file extension. That is more the domain of file explorer apps, to be able to identify file types and show them in the UI correctly. As proof, you can rename an Opus audio file downloaded from Newpipe to `.opus`, `.ogg`, and `.oga`, and it will play just fine each time (assuming your media player can recognise both: Opus the codec, and Ogg the container format. Try Poweramp, or Vanilla Music).

**Comment by Tanishq-Banyal** [comment]
My music player (musicolet) does recognise .opus as well as .ogg, just my system file explorer (Android Documents UI) doesn't recognise .opus.

I am not affected much by this as I have a script running in background that renames any .opus file to .ogg in Music folder whenever a new .opus file is detected.

**Comment by opusforlife2** [comment]
Is your LOS 16 build older than the commits I linked to? That's the only reason I can think of why your system file explorer isn't recognising them.

I'm on LOS 17.1, and the system Files app shows the file type of `.opus` files as "OGA audio", and the MIME type as "audio/ogg".

**Comment by Tanishq-Banyal** [comment]
> I'm on LOS 17.1, and the system Files app shows the file type of `.opus` files as "OGA audio", and the MIME type as "audio/ogg".

I am on LOS 16.0, and system files app shows mime type "application/octet-stream" ie. binary file

But as I said earlier :-
> I am not affected much by this as I have a script running in background that renames any .opus file to .ogg in Music folder whenever a new .opus file is detected.

**Comment by opusforlife2** [comment]
Okay. Then let's limit this issue to keeping `.ogg` for Android 9-, and `.opus` for Android 10+. This will be much easier than adding Matroska container support for Android 5 and 6.

**Comment by TacoTheDank** [comment]
> @TacoTheDank Using the `.ogg` file extension will only work for Android 7, 8, and 9, because support for it was added in 7: https://en.wikipedia.org/wiki/Opus_(audio_format)#Operating_system_support

Yep I'm aware, I just shortened it in my statement and assumed that people visiting this issue would read as such in the linked Wikipedia article. Maybe I should have clarified, lol.

**Comment by opusforlife2** [comment]
> assumed that people visiting this issue would read 

Now _there's_ a super-optimistic person. xD

**Comment by Atemu** [comment]
Quoting [rfc7845](https://datatracker.ietf.org/doc/html/rfc7845):

> The RECOMMENDED filename extension for Ogg Opus files is '.opus'.

As this is only a recommendation, I'd prefer to stick to `ogg` because OGG is a somewhat common audio container format that people know and the most widely used file extension for it is `ogg`, no matter the inner codec.

I'm also doubtful that the crappy music players people love to use can handle the `opus` extension well (despite Android 10) while I'd expect most of them to be able to handle `ogg`.

What do you think of an audio format option to save files as `webm` that gets enabled by default when the initial run happens on Android <7?  
Default with that option off would be `ogg`.

***

OT: @opusforlife2's username checks out ;D

**Comment by opusforlife2** [comment]
I don't see why we should mislabel files because people use crappy/old music apps. Instead, this could nudge them toward using updated music apps that _do_ recognise the `.opus` extension. Android 10 was released in _2019_. Any competent music app developer should have added support for it after that point.

**Comment by Atemu** [comment]
> Instead, this could nudge them toward using updated music apps that do recognise the `.opus` extension.

While the idealist in me agrees, the realist sees people that are confused about weird audio formats their favourite music player can't and blame NewPipe for it.

> Any competent music app developer should have added support for it after that point.

This implies competence among said developers which I don't think exists universally.

**Comment by opusforlife2** [comment]
I just searched, and we've never actually had anyone complain about not being able to play Opus files. (Disclaimer, there is _one_ issue, but back then Newpipe's Opus support itself was broken.)

**Comment by Tanishq-Banyal** [comment]
just let the user choose file extension in download settings

**Comment by Tanishq-Banyal** [comment]
just let the user choose file extension in download settings

**Comment by opusforlife2** [comment]
We're always conservative about introducing additional options that need to be maintained. In this case, an option isn't needed at all, because the suggested defaults make sense:

`.opus` for 10+ and `.ogg` for 5-9.

**Comment by Tanishq-Banyal** [comment]
Then finalize this ?

**Comment by opusforlife2** [comment]
What does that mean, exactly? The discussion is open to anyone who wants to add anything. Apart from that, we need a PR to implement this.

**Comment by otonoton** [comment]
> there is no such thing as a .opus file

From https://datatracker.ietf.org/doc/html/rfc7845

>    The RECOMMENDED filename extension for Ogg Opus files is '.opus'.
